File: python/1670. Design Front Middle Back Queue.py
class FrontMiddleBackQueue(object):

    def __init__(self):
        self.que = []


    def pushFront(self, val):
        """
        :type val: int
        :rtype: None
        """
        self.que.insert(0, val)


    def pushMiddle(self, val):
        """
        :type val: int
        :rtype: None
        """
        self.que.insert(self.midx(), val)


    def pushBack(self, val):
        """
        :type val: int
        :rtype: None
        """
        self.que.append(val)


    def popFront(self):
        """
        :rtype: int
        """
        if not self.que: return -1
        return self.que.pop(0)


    def popMiddle(self):
        """
        :rtype: int
        """
        if not self.que: return -1
        return self.que.pop(self.midx())


    def popBack(self):
        """
        :rtype: int
        """
        if not self.que: return -1
        return self.que.pop()

    def midx(self):
        return (len(self.que)+1)/2


if __name__ == '__main__':
    su = FrontMiddleBackQueue()
    su.pushFront(1)
    su.pushBack(2)
    su.pushMiddle(3)
    su.pushMiddle(4)
    su.popMiddle()
    su.popMiddle()
    su.popBack()
    su.popFront()
    print su.que

# Your FrontMiddleBackQueue object will be instantiated and called as such:
# obj = FrontMiddleBackQueue()
# obj.pushFront(val)
# obj.pushMiddle(val)
# obj.pushBack(val)
# param_4 = obj.popFront()
# param_5 = obj.popMiddle()
# param_6 = obj.popBack()

File: go/src/leetcode/1670. Design Front Middle Back Queue/1670. Design Front Middle Back Queue.go
/**
@author ZhengHao Lou
Date    2021/11/05
*/
package main

type FrontMiddleBackQueue struct {
	a []int
}


func Constructor() FrontMiddleBackQueue {
	return FrontMiddleBackQueue{
		a: []int{},
	}
}


func (this *FrontMiddleBackQueue) PushFront(val int)  {
	this.a = append(this.a, 0)
	copy(this.a[1:], this.a)
	this.a[0] = val
}


func (this *FrontMiddleBackQueue) PushMiddle(val int)  {
	var m = len(this.a) >> 1
	this.a = append(this.a, 0)
	copy(this.a[m+1:], this.a[m:])
	this.a[m] = val
}


func (this *FrontMiddleBackQueue) PushBack(val int)  {
	this.a = append(this.a, 0)
	this.a[len(this.a) - 1] = val
}


func (this *FrontMiddleBackQueue) PopFront() int {
	if len(this.a) == 0 {
		return -1
	}
	var val = this.a[0]
	copy(this.a, this.a[1:])
	this.a = this.a[:len(this.a) - 1]
	return val
}


func (this *FrontMiddleBackQueue) PopMiddle() int {
	if len(this.a) == 0 {
		return -1
	}
	var m = (len(this.a) - 1) >> 1
	var val = this.a[m]
	copy(this.a[m:], this.a[m+1:])
	this.a = this.a[:len(this.a) - 1]
	return val
}


func (this *FrontMiddleBackQueue) PopBack() int {
	if len(this.a) == 0 {
		return -1
	}
	var val = this.a[len(this.a) - 1]
	this.a = this.a[:len(this.a) - 1]
	return val
}


/**
 * Your FrontMiddleBackQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.PushFront(val);
 * obj.PushMiddle(val);
 * obj.PushBack(val);
 * param_4 := obj.PopFront();
 * param_5 := obj.PopMiddle();
 * param_6 := obj.PopBack();
 */